src/foam/nanos/crunch/UserCapabilityJunctionDAO.js

```diff
@@ -138,23 +162,53 @@ foam.CLASS({
 
       DAO dao = (DAO) x.get(daoKey);
       if ( dao == null ) return;
 
-      if ( dao.getOf().getId().equals((obj.getData()).getClassInfo().getId()) ) {
-        try {
-          dao.put(obj.getData());
-        } catch (Exception e) {
-          Logger logger = (Logger) x.get("logger");
-          logger.debug("Data cannot be added to " + daoKey + " for UserCapabilityJunction object : " + obj.getId() );
+      FObject objectToSave;                                                  // Identify or create data to go into dao.
+      String contextDAOFindKey = (String)capability.getContextDAOFindKey();
+
+      if ( contextDAOFindKey != null && ! contextDAOFindKey.isEmpty() ) {
+        if ( contextDAOFindKey.toLowerCase().contains("subject") ) {         // 1- Case if subject lookup
+          String[] words = foam.util.StringUtil.split(contextDAOFindKey, '.');
+          objectToSave = (FObject) x.get("subject");
+          
+          if ( objectToSave == null || words.length < 2 )
+            throw new RuntimeException("@UserCapabilityJunction capability.contextDAOFindKey not found in context. Please check capability: " + obj.getTargetId() + " and its contextDAOFindKey: " + contextDAOFindKey);
+          
+          if ( words[1].toLowerCase().equals("user") ) {
+            objectToSave = ((Subject) objectToSave).getUser();
+          } else if ( words[1].toLowerCase().equals("realuser") ) {
+            objectToSave = ((Subject) objectToSave).getRealUser();
+          }
+        } else {                                                              // 2- Case if anything other then subject specified
+          objectToSave = (FObject) x.get(contextDAOFindKey);
+
+          if ( objectToSave == null )
+            throw new RuntimeException("@UserCapabilityJunction capability.contextDAOFindKey not found in context. Please check capability: " + obj.getTargetId() + " and its contextDAOFindKey: " + contextDAOFindKey);
+        }
+      } else {
+        try {                                                                 // 3- Case where contextDAOFindKey not specified:
+          // Create new object of DAO type to copy over properties
+          objectToSave = (FObject) dao.getOf().newInstance();
+        } catch (Exception e) {                                               // 4- default case, try using ucj data directly.
+          objectToSave = (FObject) obj.getData();
         }
       }
+      objectToSave = objectToSave.fclone().copyFrom(obj.getData());           // finally copy user inputed data into objectToSave <- typed to the safest possibility from above cases
+
+      try {                                                                   // save data to dao
+        dao.put(objectToSave);
+      } catch (Exception e) {
+        Logger logger = (Logger) x.get("logger");
+        logger.warning("Data cannot be added to " + capability.getDaoKey() + " for UserCapabilityJunction object : " + obj.getId() );
```

Comment:
I don’t see any logic for updating existing objects properties here, isn’t this something that we would come across fairly frequently with expired UCJ's that are attempting to be renewed?